interfaces/prompting/patterns/variant.go

```diff
@@ -0,0 +1,426 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"regexp"
+	"strings"
+)
+
+type componentType int
+
+// Component types in order from lowest to highest precedence.
+//
+// A literal exactly matches the next non-zero number of characters, so it has
+// the highest precedence. The '?' character matches exactly one non-separator
+// character, so it has precedence over globstars; the only way a '?' can match
+// a path at the same component position as another lower-precedence component
+// is if the '?' were to occur after a '*' (and zero or more '?'s), and if this
+// were the case, then the '?' would give additional information about the
+// length of the path component matched by the '*', so it is higher precedence
+// than the separator or terminal following the '*'. Next, a separator ('/') is
+// like a literal, except that when following a '*', it precludes any more
+// information being given about the length or content of the path prior to the
+// separator (e.g. "/foo*bar/baz" clearly has precedence over "/foo*/baz"), so
+// '/' has lower precedence than literals or '?' but higher than the rest. Next,
+// terminals match when there is no path left, so they have precedence over all
+// the variable-length component types, which may match zero or more characters.
+// The next three component types relate to doublestars, which always follow a
+// '/' character, and may match zero or more characters; in order to know
+// whether a '/' is followed by a "**" or not without looking ahead in the list
+// of components, we group these components together, along with their suffix
+// when relevant: the non-terminal "/**" must be followed by a component which
+// gives more information about the matching path, so it has the highest
+// precedence of the three doublestar components; the terminal "/**/" component
+// means that the variant only matches directories, while the terminal "/**"
+// component can match files or directories, so the former has precedence over
+// the latter. Lastly, '*' has the lowest precedence, since all other component
+// types begin with more information about the length or content of the next
+// characters in the path: "/foo/**" has precedence over "/foo*" since the
+// former matches "/foo" exactly or a path in the "/foo" directory, while
+// "/foo*" matches any path which happens to begin with "/foo".
```

Comment:
Technically it matches any file or directory path what begins in `/foo`. So it would match `/foo`, `/foobar`, `/foo/`, `/foobar/`, etc. So I'd rather not use the "file path" here, unless we also use "directory path".